lib/duniverse_build.ml

```diff
@@ -0,0 +1,89 @@
+let latest_patch_release_with_flambda compiler =
+  let open Ocaml_version in
+  of_string_exn compiler |> fun ov ->
+  Releases.all_patches |>
+  List.filter (fun ov' -> with_just_major_and_minor ov' = ov) |>
+  List.sort compare |> List.rev |> function
+  | [] -> failwith ("Unknown compiler release" ^ compiler)
+  | hd::_ -> Opam.V2.name @@ with_variant hd (Some "musl+flambda") (* TODO add `Musl to ocaml-version *)
+
+let install_bin ~compiler ~repo ~tag ~bins name =
+  let open Dockerfile in
+  let alias = Printf.sprintf "%s_binary" name in
+  let base = Printf.sprintf "ocurrent/opam:debian-10-ocaml-%s" compiler in
+  let switch = latest_patch_release_with_flambda compiler in
+  let build =
+    from ~alias base @@
+    run "sudo apt-get -y install musl-tools" @@
+    run "opam switch create static %s" switch @@ (* TODO base-compilers should build this variant *)
+    run "opam install -y dune" @@
+    run "git clone --depth=1 --branch=%s https://github.com/%s.git /home/opam/src" tag repo @@
+    workdir "/home/opam/src" @@
+    run "echo \"(lang dune 2.0)\" > dune-workspace" @@
+    run "echo \"(env (_ (flags -cclib -static)))\" >> dune-workspace" @@
+    run "opam pin add -n ." @@
+    run "opam depext %s" name @@
+    run "opam install --deps-only %s" name @@
+    run "opam exec -- dune build @install" @@@
+    (List.map (fun (bin, target) ->
+       run "sudo install -m 0755 -o root _build/install/default/%s %s" bin target) bins)
+  in
+  let use = copy ~from:alias ~src:(List.map snd bins) ~dst:"/usr/bin/" () in
+  build, use
+
+let install_platform ~compiler =
+  let duniverse = install_bin ~compiler
+    ~repo:"ocamllabs/duniverse" ~tag:"dev-ocaml-ci-3"
+    ~bins:["bin/duniverse", "/usr/bin/duniverse"] "duniverse" in
+  let dune = install_bin ~compiler
+    ~repo:"ocaml/dune" ~tag:"2.5.0"
+    ~bins:["bin/dune", "/usr/bin/dune"] "dune" in
+  let odoc = install_bin ~compiler
+    ~repo:"ocaml/odoc" ~tag:"1.5.0"
+    ~bins:["bin/odoc", "/usr/bin/odoc"] "odoc" in
+  let ocamlformat = install_bin ~compiler
+    ~repo:"ocaml-ppx/ocamlformat" ~tag:"0.14.0"
```

Comment:
Hm, this is probably buggy.  We need to install the precise ocamlformat version required by the project, and that is found in the analysis phase.